crates/ide-completion/src/context/analysis.rs

```diff
@@ -719,6 +719,80 @@ fn classify_name_ref(
+                                    }
+                                    hir::ModuleDef::Variant(variant) => {
+                                        variant.parent_enum(sema.db).source(sema.db)?.value.generic_param_list()
+                                    }
+                                    hir::ModuleDef::Trait(trait_) => {
+                                        trait_.source(sema.db)?.value.generic_param_list()
+                                    }
+                                    hir::ModuleDef::TraitAlias(trait_) => {
+                                        trait_.source(sema.db)?.value.generic_param_list()
+                                    }
+                                    hir::ModuleDef::TypeAlias(ty_) => {
+                                        ty_.source(sema.db)?.value.generic_param_list()
+                                    }
+                                    _ => None,
+                                },
+                                _ => None,
+                            }
+                        },
+                        ast::MethodCallExpr(call) => {
+                            let func = sema.resolve_method_call(&call)?;
+                            func.source(sema.db)?.value.generic_param_list()
+                        },
+                        ast::AssocTypeArg(arg) => {
+                            let trait_ = ast::PathSegment::cast(arg.syntax().parent()?.parent()?)?;
+                            match sema.resolve_path(&trait_.parent_path().top_path())? {
+                                hir::PathResolution::Def(def) => match def {
+                                    hir::ModuleDef::Trait(trait_) => {
+                                        let trait_items = trait_.items(sema.db);
+                                        let assoc_ty = trait_items.iter().find_map(|item| match item {
+                                            hir::AssocItem::TypeAlias(assoc_ty) => {
+                                                (assoc_ty.name(sema.db).as_str()? == arg.name_ref()?.text())
+                                                    .then_some(assoc_ty)
+                                            },
+                                            _ => None,
+                                        })?;
+                                        assoc_ty.source(sema.db)?.value.generic_param_list()
+                                    }
+                                    _ => None,
+                                },
+                                _ => None,
+                            }
+                        },
+                        _ => None,
+                    }
+                }?;
+                // Determine the index of the parameter in the `GenericArgList`
+                // (subtract 1 because `siblings` includes the node itself)
+                let param_idx = arg.syntax().siblings(Direction::Prev).count() - 1;
+                params.generic_params().nth(param_idx)
```

Comment:
There is also a problem here, you can omit lifetime generic arguments which this code does not account for, that is
```rs
struct S<'a, 'b, T, const C: usize>;

S::<$0, _>
```
here your code will match this up with the lifetime argument which is not wrong per se, it is a lifetime argument position (though in this case you cannot specify a lifetime as its a late bound paramter), *but* it is also a position for a type argument. So for simplicities sake we should ignore all leading lifetime arguments when calculating our position, and all leading lifetime parameters when calculating the "target" position.

Reply:
Added logic for this that ignores lifetime params unless a lifetime arg is provided explicitly, in which case it will match lifetime argument positions to lifetime parameters:
https://github.com/max-heller/rust-analyzer/blob/bed1114b8ba258acb243c938e7b2915424426474/crates/ide-completion/src/tests/type_pos.rs#L937-L974